modules/internal/DefaultRectangular.chpl

```diff
@@ -1352,20 +1352,20 @@ module DefaultRectangular {
     where rank == 1 do
       return dsiAccess(ind);
 
-    inline proc dsiAccess(ind : rank*idxType) ref {
+    inline proc dsiAccess(const in ind : rank*idxType) ref {
       // Note: bounds checking occurs in ChapelArray for this type.
       var dataInd = getDataIndex(ind);
       return theData(dataInd);
     }
 
-    inline proc dsiAccess(ind : rank*idxType)
+    inline proc dsiAccess(const in ind : rank*idxType)
     where shouldReturnRvalueByValue(eltType) {
       // Note: bounds checking occurs in ChapelArray for this type.
       var dataInd = getDataIndex(ind);
       return theData(dataInd);
     }
 
-    inline proc dsiAccess(ind : rank*idxType) const ref {
+    inline proc dsiAccess(const in ind : rank*idxType) const ref {
```

Comment:
The title of this PR says `const ref`, but these are `const in`.  Is the title just outdated?

@vasslitvinov / @dlongnecke-cray : `const ref` improving performance would make sense to me since presumably some temps are removed (though I could also imagine it hurting performance due to dereferences).  I'm not understanding how / why / when `const in` would be different than the default intent for tuples here, though—i.e., is the default for tuples not already `const in`?  (`const ref` would be my second guess...).

Is this clear to you?

Reply:
For me, the notion was the opposite: if `dsiAccess` takes a `const ref` to a tuple, then it seems like it's going to have to coerce if it's passing to either blank intent or in intent. My notion was that the cost of the deref + the coercion is more painful than just copying in the tuple of indices. If the whole stack of calls from array.this down uses `const in` then it would just be bitcopies all the way.

You're right that the default intent for `2*int` effectively stores each `int` as `in`, but I'm not certain that the blank intent tuple _type_ for `2*int` is equivalent to the one produced for `const in 2*int`. If it's not, perhaps it should be, but it still seems more logically consistent to me to be explicitly bitcopying our array indices around anyway. So `const in` seems more honest here in addition to avoiding any unintentional conversions.

Reply:
The reason I imagined `const in` to be the default intent for this tuple is that I thought we'd been taking a philosophy of having passing 1-tuples of `t` as having similar behavior as simply passing an argument of type `t`.  So since `int`s, `enum`s, `bool`s are passed by `const in`, I was expecting tuples of the same to be as well.

If that was wrong, I would've guessed `const ref` to be the default since it's the only other `const` option, and I'd be surprised if a non-`const` default was used...

If I'm understanding you, David, you're saying that you'd expect that the behavior was `const in`-like, but that something about the implementation isn't literally the same as `const in` and therefore causing a difference in behavior / performance?  That seems unsettling to me...

Reply:
> If I'm understanding you, David, you're saying that you'd expect that the behavior was const in-like, but that something about the implementation isn't literally the same as const in and therefore causing a difference in behavior / performance?

Right, I'm not sure about this and could be wrong. But Vass and I definitely observed the compiler doing more coercions which involves unpacking and repacking each element.

I agree it's concerning that there may potentially be some unnecessary coercions going on (and perhaps we can minimize those going forward). I still think that `const in` is the right thing to do for these methods and IMO we should consider adjusting some more of the helpers for DefaultRectangular to use `const in` in the future.

What I don't understand is that I'm not sure any other methods in this stack currently use `const in` besides `dsiAccess`, so I'm not sure why this fixes the performance degradation.

Reply:
@vasslitvinov : Do you have any insights here as to why/how the compiler could have been behaving/performing differently when using blank intents compared to using `const in` or `const ref`?  If not, should we open an issue to capture and try to resolve this mystery?

Reply:
`const ref` brought the concrete intents back to what they were de-facto implicitly prior to #21665. Still there are plenty of follow-up questions to resolve. So...

I opened #21901 to track them.